plugins/woocommerce-admin/client/marketing/data-multichannel/selectors.ts

```diff
@@ -23,6 +23,15 @@ export const getCampaigns = ( state: State, page: number, perPage: number ) => {
 	return state.campaigns.pages[ key ] || null;
 };
 
+/**
+ * Get the meta of campaigns from state.
+ *
+ * @param state State passed in from the data store.
+ */
+export const getCampaignsMeta = ( state: State ) => {
+	return state.campaigns.meta;
+};
+
```

Comment:
Maybe I was too focused on maintaining the stability of returning references.

> I think a better way is to return the `meta` / `meta.total` in the `getCampaigns` selector, so that when we call `getCampaigns`, we would get the campaigns page and also the total in one selector, and the resolver would also get called.

Considering that `getCampaigns` doesn't actually have rendering performance issues, and that even `useCampaigns` itself doesn't consider the stability of returning references, it should be fine for now.

if necessary in the future, wrapping `getCampaigns` with `rememo` or referring to something like [this resolver](https://github.com/woocommerce/woocommerce/blob/54b980c2d2512ff37c5fee3c869e325a8dce31f3/packages/js/data/src/crud/resolvers.ts#L107-L146) should make references stable.

Revised in https://github.com/woocommerce/woocommerce/pull/41211/commits/17c624ad5cbcea5671b3364da44eb1b9c40ff204.

Reply:
I did some digging and I think I get what you mean now. For example, in `useCampaigns`, we have this line:

```js
data: campaignsPage?.data?.map( convert ),
```

Even though the campaigns data in the data store did not change, because we are using `.map( convert )`, it returns a new array with the same data every time, so the reference is not stable and it causes unnecessary re-renders.

And I think the same issue happened for the `useCampaigns` code before this PR, where we were returning a new object for the `meta` field:

```js
				meta: {
					total: campaignsState.total,
				},
```

In this PR, because now we are returning the `meta` field directly from the data store without creating a new object, so we do not have this issue with the `meta` field.

Is there anything else that I'm missing? 🤔 

Reference: https://jardasn.blog/2021/01/22/some-best-practices-for-using-useselect-from-wordpress-data/

Reply:
Yes, it's recommended to keep the references shallowly equal to the previous one when using `useSelect`, but the `data` property still doesn't meet that after this PR.

For the `getCampaigns` selector, it shares the same concept of redux that avoids creating derived values that are new references. So there are a number of WP and Woo practices that use a memoization library, which is `rememo`, to achieve this.